libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/WaymarkOverlay.kt

```diff
@@ -0,0 +1,170 @@
+    private val screenPos = Point()
+    private val bitmapTransform = Matrix()
+    private val bitmapPaint = Paint()
+
+    private val shareDrawable: Drawable
+    private val commentDrawable: Drawable
+
+    private val waymarkers: MutableList<OverlayItem>
+
+    init {
+        val context = mapView.context
+
+        val res = context.resources
+        greenWisp = ResourcesCompat.getDrawable(res, R.drawable.green_wisp, null)
+        orangeWisp = ResourcesCompat.getDrawable(res, R.drawable.orange_wisp, null)
+        redWisp = ResourcesCompat.getDrawable(res, R.drawable.red_wisp, null)
+
+        shareDrawable = IconicsDrawable(context)
+                .icon(GoogleMaterial.Icon.gmd_share)
+                .color(Theme.lowlightColorInverse(context))
+                .sizeDp(24)
+        commentDrawable = IconicsDrawable(context)
+                .icon(GoogleMaterial.Icon.gmd_comment)
+                .color(Theme.lowlightColorInverse(context))
+                .sizeDp(24)
+
+        waymarkers = ArrayList()
+    }
+
+    private fun setWaypoints(waypoints: Waypoints) {
+        resetWaypoints()
+
+        for (waypoint in waypoints) {
+            addWaypoint(waypoint)
+        }
+    }
+
+    private fun resetWaypoints() {
+        waymarkers.clear()
+    }
+
+    fun waymarkersCount(): Int {
+        return waymarkers.size
+    }
+
+    fun waypoints(): Waypoints {
+        val p = ArrayList<IGeoPoint>()
+        for (o in waymarkers)
+            p.add(o.point)
+        return Waypoints(p)
+    }
+
+    fun finish(): IGeoPoint {
+        return waymarkers[waymarkersCount() - 1].point
+    }
+
+    fun addWaypoint(point: IGeoPoint?) {
+        if (point == null)
+            return
+        when (waymarkersCount()) {
+            0 -> waymarkers.add(addMarker(point, "start", greenWisp))
+            1 -> waymarkers.add(addMarker(point, "finish", redWisp))
+            else -> {
+                val prevFinished = finish()
+                waymarkers.removeAt(waymarkersCount() - 1)
+                waymarkers.add(addMarker(prevFinished, "waypoint", orangeWisp))
+                waymarkers.add(addMarker(point, "finish", redWisp))
+            }
+        }
+    }
+
+    fun removeWaypoint() {
+        when (waymarkersCount()) {
+            0 -> {
+            }
+            1, 2 -> waymarkers.removeAt(waymarkersCount() - 1)
+            else -> {
+                waymarkers.removeAt(waymarkersCount() - 1)
+                val prevFinished = finish()
+                waymarkers.removeAt(waymarkersCount() - 1)
+                waymarkers.add(addMarker(prevFinished, "finish", redWisp))
+            }
+        }
+    }
+
+    private fun addMarker(point: IGeoPoint, label: String, icon: Drawable?): OverlayItem {
+        val marker = OverlayItem(label, label, GeoPoint(point.latitude, point.longitude))
+        marker.setMarker(icon)
+        marker.markerHotspot = OverlayItem.HotspotPlace.BOTTOM_CENTER
+        return marker
```

Comment:
Not essential, but if you want to make this more concise, you could change it to:

```kotlin
        return OverlayItem(label, label, GeoPoint(point.latitude, point.longitude)).apply {
            setMarker(icon)
            markerHotspot = OverlayItem.HotspotPlace.BOTTOM_CENTER
        }
```